### PR TITLE
[Model Zoo] Sketch

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by VoVAllen** [comment]
Sample at examples/pytorch/model_zoo/classification.py

File: examples/pytorch/model_zoo/classification.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+
+from utils import Meter, EarlyStopping, collate_molgraphs
+
+def main():
+    device = "cuda" if torch.cuda.is_available() else "cpu"
+    batch_size = 128
+    learning_rate = 0.001
+    num_epochs = 100
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:15]: Unless you've re-tuned the model to match the performance and updated the pretrained model, I'd suggest changing this back to my old configurations to ensure reproducible performance. Also these should be arguments.

File: examples/pytorch/model_zoo/classification.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    # Interchangable to other Dataset
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:17]: Interchangeable with other datasets

File: examples/pytorch/model_zoo/classification.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    trainset, valset, testset = split_dataset(dataset, [0.8, 0.1, 0.1])
+    train_loader = DataLoader(
+        trainset, batch_size=batch_size, collate_fn=collate_molgraphs)
+    val_loader = DataLoader(
+        valset, batch_size=batch_size, collate_fn=collate_molgraphs)
+    test_loader = DataLoader(
+        testset, batch_size=batch_size, collate_fn=collate_molgraphs)
+
+    # Interchangable to other model in model zoo
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:30]: Same

File: examples/pytorch/model_zoo/classification.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    model = model_zoo.chem.GCNClassifier(74, [64, 64], dataset.n_tasks)
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:31]: Try to use dictionary for configuration and you can do something like `model_zoo.chem.GCNClassifier(**configuration)`.

File: examples/pytorch/model_zoo/classification.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    model.to(device)
+    loss_criterion = BCEWithLogitsLoss(pos_weight=torch.tensor(
+        dataset.task_pos_weights).to(device), reduction='none')
+    optimizer = Adam(model.parameters(), lr=learning_rate)
+
+    stopper = EarlyStopping(patience=10)
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:38]: No need to include EarlyStopping for the time being.

File: examples/pytorch/model_zoo/classification.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    for epoch in range(num_epochs):
+        model.train()
+        print('Start training')
+        train_meter = Meter()
+        model.train()
+        for batch_id, batch_data in enumerate(train_loader):
+            smiles, bg, labels, mask = batch_data
+            atom_feats = bg.ndata.pop(atom_data_field)
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:47]: We can probably move this line into `collate_molgraphs`.

File: examples/pytorch/model_zoo/classification.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
+from dgl.data import Tox21
+from dgl.data.utils import split_dataset
+from dgl import model_zoo
+import torch
+from torch.nn import BCEWithLogitsLoss
+from torch.optim import Adam
+from torch.utils.data import DataLoader
+
+from utils import Meter, EarlyStopping, collate_molgraphs
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:9]: Why do you want to call this staff `Meter`? Are there any other well recognized examples using similar naming?

File: python/dgl/data/chem/utils.py [hunk 8]
```diff
@@ -141,7 +142,7 @@ def __call__(self, mol):
         return {self.atom_data_field: atom_features}
 
 
-def smile2graph(smile, add_self_loop=False, atom_featurizer=None, bond_featurizer=None):
+def smile2graph(smile, add_self_loop=False, atom_featurizer=DefaultAtomFeaturizer(), bond_featurizer=None):
```
Comment (mufeili) [inline python/dgl/data/chem/utils.py:145]: If you change the default arg, you need to also change the related documentation.

File: examples/pytorch/model_zoo/utils.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        model.load_state_dict(torch.load(self.filename))
+
+
+def collate_molgraphs(data):
+    """Batching a list of datapoints for dataloader
+    Parameters
+    ----------
+    data : list of 3-tuples
+        Each tuple is for a single datapoint, consisting of
+        a DGLGraph, all-task labels and all-task weights
```
Comment (mufeili) [inline examples/pytorch/model_zoo/utils.py:74]: Update the documentation here as we now include SMILES.

File: examples/pytorch/model_zoo/classification.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        dataset.task_pos_weights).to(device), reduction='none')
+    optimizer = Adam(model.parameters(), lr=learning_rate)
+
+    stopper = EarlyStopping(patience=10)
+
+    for epoch in range(num_epochs):
+        model.train()
+        print('Start training')
+        train_meter = Meter()
+        model.train()
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:44]: Duplicate `model.train()`.

File: python/dgl/model_zoo/chem/pretrain.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+from ...data.utils import _get_dgl_url, download
+
+
+def load_pretrained(model_name):
+    if model_name == "GCN_Tox21":
+        print('Loading pretrained model...')
+        url_to_pretrained = _get_dgl_url('pre_trained/gcn_tox21.pth')
+        local_pretrained_path = 'pre_trained.pth'
+        download(url_to_pretrained, path=local_pretrained_path)
+        model = GCNClassifier(74, [64, 64], 12, 64)
```
Comment (mufeili) [inline python/dgl/model_zoo/chem/pretrain.py:12]: Add names of arguments for clarity

File: python/dgl/model_zoo/chem/pretrain.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+        checkpoint = torch.load(local_pretrained_path)
+        model.load_state_dict(checkpoint['model_state_dict'])
+        return model
```
Comment (mufeili) [inline python/dgl/model_zoo/chem/pretrain.py:15]: Raise an error if `model_name` is unexpected.

File: examples/pytorch/model_zoo/utils.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    -------
+    smiles : list
+        List of smiles
+    bg : BatchedDGLGraph
+        Batched DGLGraphs
+    labels : Tensor of dtype float32 and shape (B, 12)
+        Batched datapoint labels. B is len(data) and
+        12 is the number of total tasks.
+    weights : Tensor of dtype float32 and shape (B, 12)
+        Batched datapoint weights.
```
Comment (mufeili) [inline examples/pytorch/model_zoo/utils.py:85]: If this function is going to be re-used, the number of labels is not restricted to be "12".

File: examples/pytorch/model_zoo/utils.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+        for task in range(n_tasks):
+            task_w = mask[:, task]
+            task_y_true = y_true[:, task][task_w != 0].cpu().numpy()
+            task_y_pred = y_pred[:, task][task_w != 0].cpu().detach().numpy()
+            total_score += roc_auc_score(task_y_true, task_y_pred)
+        return total_score / n_tasks
+
+
+class EarlyStopping:
+    def __init__(self, patience=10, filename="es_checkpoint.pt"):
```
Comment (mufeili) [inline examples/pytorch/model_zoo/utils.py:35]: To be consistent, I'd suggest changing ``es_checkpoint.pt`` to ``es_checkpoint.pth`` unless you want to change the pre-trained file.

File: examples/pytorch/model_zoo/classification.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+            # Mask non-existing labels
+            loss = (loss_criterion(logits, labels)
+                    * (mask != 0).float()).mean()
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+            print('epoch {:d}/{:d}, batch {:d}/{:d}, loss {:.4f}'.format(
+                epoch + 1, num_epochs, batch_id + 1, len(train_loader), loss.item()))
+            train_meter.update(logits, labels, mask)
+        train_roc_auc = train_meter.roc_auc_averaged_over_tasks()
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:59]: I would still suggest wrapping `L41-59` to make the main body shorter.

File: examples/pytorch/model_zoo/classification.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+        model.eval()
+        with torch.no_grad():
+            for batch_id, batch_data in enumerate(val_loader):
+                smiles, bg, labels, mask = batch_data
+                atom_feats = bg.ndata.pop(atom_data_field)
+                atom_feats, labels = atom_feats.to(device), labels.to(device)
+                logits = model(atom_feats, bg)
+                val_meter.update(logits, labels, mask)
+        
+        val_roc_auc = val_meter.roc_auc_averaged_over_tasks()
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:72]: Same.

File: examples/pytorch/model_zoo/classification.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+        print('epoch {:d}/{:d}, validation roc-auc score {:.4f}, best validation roc-auc score {:.4f}'.format(
+            epoch + 1, num_epochs, val_roc_auc, stopper.best_score))
+
+    test_meter = Meter()
+    for batch_id, batch_data in enumerate(test_loader):
+        smiles, bg, labels, mask = batch_data
+        atom_feats = bg.ndata.pop(atom_data_field)
+        atom_feats, labels = atom_feats.to(device), labels.to(device)
+        logits = model(atom_feats, bg)
+        test_meter.update(logits, labels, mask)
```
Comment (mufeili) [inline examples/pytorch/model_zoo/classification.py:85]: Same.

File: examples/pytorch/model_zoo/inference_pretrained.ipynb [hunk 18]
```diff
@@ -0,0 +1,304 @@
+{
```
Comment (mufeili) [inline examples/pytorch/model_zoo/inference_pretrained.ipynb:1]: Will postpone review for jupyter notebook until it's mature

File: examples/pytorch/model_zoo/classification.py (hunk 1 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/classification.py:15]: This is an use case demonstration, user should not use it with arguments. 
Reproducing doesn't work with gpu, I don't want to introduce the complexity here.

File: examples/pytorch/model_zoo/classification.py (hunk 2 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/classification.py:17]: Done

File: examples/pytorch/model_zoo/classification.py (hunk 3 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/classification.py:30]: Done

File: examples/pytorch/model_zoo/classification.py (hunk 7 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/classification.py:9]: https://github.com/pytorch/examples/blob/master/imagenet/main.py#L354

File: examples/pytorch/model_zoo/classification.py (hunk 10 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/classification.py:44]: Removed

File: examples/pytorch/model_zoo/utils.py (hunk 14 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/utils.py:35]: Done

File: examples/pytorch/model_zoo/classification.py (hunk 15 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/classification.py:59]: I prefer not enclosing them into another function since they are short enough. The main reason is that function will take many things as argument. I would have no idea what's happening inside `run_a_train_epoch` since it's taking everything into it. Current ~80 lines I think is a proper length to read.

File: examples/pytorch/model_zoo/classification.py (hunk 6 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/classification.py:47]: I think it's fine to put it here, making it consistent with `dataset.__getitem__`

File: examples/pytorch/model_zoo/utils.py (hunk 9 above)
Comment (VoVAllen) [reply examples/pytorch/model_zoo/utils.py:74]: Done

File: python/dgl/data/chem/utils.py (hunk 8 above)
Comment (VoVAllen) [reply python/dgl/data/chem/utils.py:145]: Done

File: python/dgl/model_zoo/chem/pretrain.py (hunk 12 above)
Comment (VoVAllen) [reply python/dgl/model_zoo/chem/pretrain.py:15]: Done.
